File: php/lc00001-lc00100/lc00005/LongestPalindromicSubstringTest.php
<?php

namespace leetcode\lc00005;

use PHPUnit\Framework\TestCase;

class LongestPalindromicSubstringTest extends TestCase
{
    private $cases = [
        ["s" => "babad", "expected" => "bab"],
        ["s" => "cbbd", "expected" => "bb"],
        ["s" => "civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth", "expected" => "ranynar"],
        ["s" => "aacabdkacaa", "expected" => "aca"],
        ["s" => "edllaaaaallf", "expected" => "llaaaaall"],
    ];

    public function testLongestPalindrome()
    {
        $obj = new LongestPalindromicSubstring();
        for ($i = 0; $i < count($this->cases); $i++) {
            $this->assertEquals($this->cases[$i]['expected'], $obj->longestPalindrome($this->cases[$i]['s']));
        }
    }

    public function testManacher()
    {
        $obj = new LongestPalindromicSubstring();
        for ($i = 0; $i < count($this->cases); $i++) {
            $this->assertEquals($this->cases[$i]['expected'], $obj->manacher($this->cases[$i]['s']));
        }
    }
}

File: golang/lc00001-lc00100/lc00005/longest_palindromic_substring_test.go
package lc00005

import (
	"testing"
)

type caseType struct {
	s        string
	excepted string
}

var cases = []caseType{
	{"abcd", "a"},
	{"babad", "bab"},
	{"cbbd", "bb"},
	{"civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth", "ranynar"},
	{"aacabdkacaa", "aca"},
	{"edllaaaaallf", "llaaaaall"},
}

func TestLongestPalindrome(t *testing.T) {
	for _, c := range cases {
		result := longestPalindrome(c.s)
		if result != c.excepted {
			t.Fatalf("longestPalindrome function failed, s: %s, execpted: %s, result: %s", c.s, c.excepted, result)
		}
	}
}

func TestCenterSpread(t *testing.T) {
	for _, c := range cases {
		result := centerSpread(c.s)
		if result != c.excepted {
			t.Fatalf("centerSpread function failed, s: %s, execpted: %s, result: %s", c.s, c.excepted, result)
		}
	}
}

func TestManacher(t *testing.T) {
	for _, c := range cases {
		result := manacher(c.s)
		if result != c.excepted {
			t.Fatalf("manacher function failed, s: %s, execpted: %s, result: %s", c.s, c.excepted, result)
		}
	}
}

func TestDynamic(t *testing.T) {
	for _, c := range cases {
		result := dynamic(c.s)
		if result != c.excepted {
			t.Fatalf("dynamic function failed, s: %s, execpted: %s, result: %s", c.s, c.excepted, result)
		}
	}
}
